ArcGISRuntimeSDKQt_CppSamples/Maps/DownloadPreplannedMap/DownloadPreplannedMap.cpp

```diff
@@ -0,0 +1,226 @@
+// [Legal]
+// Copyright 2019 Esri.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// [Legal]
+
+#include "DownloadPreplannedMap.h"
+
+#include "Map.h"
+#include "MapQuickView.h"
+#include "OfflineMapTask.h"
+#include "PreplannedMapArea.h"
+#include "PreplannedMapAreaListModel.h"
+#include "DownloadPreplannedOfflineMapJob.h"
+#include "DownloadPreplannedOfflineMapResult.h"
+#include "SimpleLineSymbol.h"
+#include "SimpleRenderer.h"
+#include "Graphic.h"
+
+using namespace Esri::ArcGISRuntime;
+
+DownloadPreplannedMap::DownloadPreplannedMap(QObject* parent /* = nullptr */):
+  QObject(parent),
+  m_portalItem(new PortalItem("acc027394bc84c2fb04d1ed317aac674", this)),
+  m_lineSymbol(new SimpleLineSymbol(SimpleLineSymbolStyle::Solid, QColor(Qt::red), 5, this)),
+  m_graphicParent(new QObject())
+{
+  m_busy = true;
+  emit busyChanged();
+
+  m_graphicsOverlay = new GraphicsOverlay(this);
+  m_graphicsOverlay->setRenderer(new SimpleRenderer(m_lineSymbol, this));
+
+  connect(m_portalItem, &PortalItem::doneLoading, this, [this] ()
+  {
+    m_offlineMapTask = new OfflineMapTask(m_map, this);
+    m_mapView->graphicsOverlays()->append(m_graphicsOverlay);
+
+    connect(m_offlineMapTask, &OfflineMapTask::doneLoading, this, [this] ()
+    {
+      connect(m_offlineMapTask, &OfflineMapTask::preplannedMapAreasCompleted, this, &DownloadPreplannedMap::loadPreplannedMapAreas);
+      m_offlineMapTask->preplannedMapAreas();
+    });
+
+    m_offlineMapTask->load();
+  });
+
+  m_map = new Map(m_portalItem, this);
+}
+
+DownloadPreplannedMap::~DownloadPreplannedMap() = default;
+
+void DownloadPreplannedMap::init()
+{
+  // Register the map view for QML
+  qmlRegisterType<MapQuickView>("Esri.Samples", 1, 0, "MapView");
+  qmlRegisterType<DownloadPreplannedMap>("Esri.Samples", 1, 0, "DownloadPreplannedMapSample");
+}
+
+MapQuickView* DownloadPreplannedMap::mapView() const
+{
+  return m_mapView;
+}
+
+// Set the view (created in QML)
+void DownloadPreplannedMap::setMapView(MapQuickView* mapView)
+{
+  if (!mapView || mapView == m_mapView)
+    return;
+
+  m_mapView = mapView;
+  m_mapView->setMap(m_map);
+
+  emit mapViewChanged();
+}
+
+void DownloadPreplannedMap::downloadMapArea(int index)
+{
+  if (!m_offlineMapTask)
+    return;
+
+  if (m_offlineMapTask->preplannedMapAreaList()->isEmpty())
+    return;
+
+  if (m_offlineMapTask->preplannedMapAreaList()->at(index)->loadStatus() != LoadStatus::Loaded)
```

Comment:
Probably better to get the area so you don't have to keep fetching it from the  model:

```suggestion
  PreplannedMapArea* mapArea = m_offlineMapTask->preplannedMapAreaList()->at(index);
  if (!mapArea || mapArea->loadStatus() != LoadStatus::Loaded)
```